### PR TITLE
refactor: Rewrote code to increase efficiency and readability.

### PR DESCRIPTION
**Pull request by JosephDiPalma**
## Changes in this Release
- Use Python [multiprocessing](https://docs.python.org/3.6/library/multiprocessing.html) to speed up patch creation.
- Replace [os](https://docs.python.org/3.6/library/os.html) with [pathlib](https://docs.python.org/3.6/library/pathlib.html) module.
- Added [argparse](https://docs.python.org/3.6/library/argparse.html) support.
- Rewrote patch processing code using vector operations.
- Fully documented all functions and methods with docstrings and type hints.
- Compute image means and standard deviations for normalization online.
- Increased efficiency of code by rewriting inefficient code.

### CODE REVIEW

**Comment by jasonwei20** [comment]
Hey Joseph, thank you for the PR. At this point, it looks like every single file was rewritten, with almost 2k lines. 

Could you outline in a comment all the major changes that you made to the code, so that previous users can see what changed?
What are the specific increases in efficiency?
Ideally this should have been in many smaller PRs, so that a single blocking change doesn't block unrelated modifications.
What test cases have you run? Since many projects in our lab use this code, I would need to see an end-to-end test with a folder of images work (just show me the logs and the final accuracy is reasonable).
At this point, do you want to be the maintainer of this repository?

CC @saeedhassanpour

File: code/3_train.py [hunk 1]
```diff
@@ -1,21 +1,6 @@
...
-                 resume_checkpoint_path=config.resume_checkpoint_path,
-                 save_interval=config.save_interval,
-                 checkpoints_folder=config.checkpoints_folder,
-                 pretrain=config.pretrain,
-                 log_csv=config.log_csv)
+from utils_model import train_resnet
+
+# Training the ResNet.
+print("+++++ Running 3_train.py +++++")
+train_resnet()
```
Comment (jasonwei20) [inline code/3_train.py]: I don't think that these parameters should be taken away. 

It's nice to see what the exact inputs are for this function.

WDYT?

File: code/4_test.py [hunk 2]
```diff
@@ -1,20 +1,16 @@
...
+# Run the ResNet on the generated patches.
+print("+++++ Running 4_test.py +++++")
+print("----- Finding validation patch predictions -----")
+# Validation patches.
+get_predictions(patches_eval_folder=config.args.patches_eval_val,
+                output_folder=config.args.preds_val)
+print("----- Finished finding validation patch predictions -----")
+print("----- Finding test patch predictions -----")
+# Test patches.
+get_predictions(patches_eval_folder=config.args.patches_eval_test,
```
Comment (jasonwei20) [inline code/4_test.py]: why did you take away these parameters

File: code/compute_stats.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+"""
+DeepSlide
+Computes the image statistics for normalization.
+
+Authors: Naofumi Tomita
```
Comment (jasonwei20) [inline code/compute_stats.py]: Did naofumi author this or did you?

File: code/config.py [hunk 4]
```diff
@@ -1,120 +1,360 @@
...
+# n/*.jpg with the images in class n.
+parser.add_argument(
+    "--all-wsi",
+    type=Path,
+    default=Path("all_wsi"),
+    help="Location of the WSI organized in subfolders by class")
+# For splitting into validation set.
+parser.add_argument("--val-wsi-per-class",
+                    type=int,
+                    default=3,
```
Comment (jasonwei20) [inline code/config.py]: why should this be 3

File: code/config.py [hunk 5]
```diff
@@ -1,120 +1,360 @@
...
 
-# folders of patches by wsi in test set, used for getting test acc on wsi level
-patches_eval_test = 'patches_eval_test'
+########################################
+#               TRAINING               #
+########################################
+# Model hyperparameters.
+parser.add_argument("--num-epochs",
+                    type=int,
+                    default=10,
```
Comment (jasonwei20) [inline code/config.py]: seems small to me

File: code/run_all.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
+import config
```
Comment (jasonwei20) [inline code/run_all.py]: i don't get why this exists if the `.sh` file is there

File: code/compute_stats.py (hunk 3 above)
Comment (JosephDiPalma) [reply code/compute_stats.py]: This was done by Naofumi using code from the PyTorch forums.  I added it in so we didn't have to rely on hard-coding image means and standard deviations each time we run.

File: code/config.py (hunk 4 above)
Comment (JosephDiPalma) [reply code/config.py]: This was 3 since the dataset I ran it on was relatively small.  It should probably be changed back to what it was initially.

File: code/config.py (hunk 5 above)
Comment (JosephDiPalma) [reply code/config.py]: Once again, this was because I was testing.  I should changed it back.

**Comment by JosephDiPalma** [comment]
@jasonwei20 I'm currently working on rewriting the README.  It should be finished today.

File: code/3_train.py (hunk 1 above)
Comment (JosephDiPalma) [reply code/3_train.py]: That was something I wasn't entirely sure about.  I agree that it's nice to see what's being passed in.

**Comment by JosephDiPalma** [comment]
@jasonwei20 I tested each function of the code individually to confirm the behavior is identical to the original one.  I then completed test runs for all hyperparameter combinations to verify the correctness.

**Comment by ntomita** [comment]
@jasonwei20 I asked @JosephDiPalma to run model on existing dataset for further verification. Also I agree that default value shouldn't be changed. I'll review this commit in a few days. Let me know if you have any other concern.

File: code/compute_stats.py (hunk 3 above)
Comment (jasonwei2020) [reply code/compute_stats.py]: LGTM

File: code/7_final_test.py [hunk 7]
```diff
@@ -1,20 +1,21 @@
...
-                       conf_thresholds=best_thresholds)
-
-print_final_test_results(labels_csv=config.labels_test,
-                         inference_folder=config.inference_test)
+from utils_evaluation import (find_best_acc_and_thresh, output_all_predictions,
+                              print_final_test_results)
+import config
+
+# Running the code on the testing set.
+print("+++++ Running 7_final_test.py +++++")
```
Comment (jasonwei20) [inline code/7_final_test.py]: This and the other print function with dashes could be wrapped in another function.

File: code/compute_stats.py (hunk 3 above)
Comment (jasonwei20) [reply code/compute_stats.py]: Shouldn't this be calculated on the patch-level, not the WSI level? Because WSI is mostly whitespace.

File: code/config.py [hunk 8]
```diff
@@ -1,120 +1,360 @@
...
-# This file contains all hyperparameters for the entire repository.
+Authors: Jason Wei, Behnaz Abdollahi, Saeed Hassanpour
 
-import utils
-from utils import *
+Last Modified: November 30, 2019 (Joseph DiPalma)
+"""
+
+import argparse
+from pathlib import Path
```
Comment (jasonwei20) [inline code/config.py]: Can you explain why this is better than just a string? 

Should this and argparse be imported in utils?

File: code/config.py [hunk 9]
```diff
@@ -1,120 +1,360 @@
...
-# for visualization
-colors = ['red', 'white', 'blue', 'green', 'purple', 'orange', 'black', 'pink', 'yellow'] 	# this order is the same order as your sorted classes
-vis_train = 'vis_train'
-vis_val = 'vis_val'
-vis_test = 'vis_test'
+# For visualization.
+# This order is the same order as your sorted classes.
+colors = [
+    "red", "white", "blue", "green", "purple", "orange", "black", "pink",
+    "yellow"
```
Comment (jasonwei20) [inline code/config.py]: why is this on a separate line

File: code/utils_evaluation.py [hunk 10]
```diff
@@ -1,119 +1,159 @@
...
+    with patches_pred_file.open(mode="r") as patches_pred:
+
+        patches_pred_lines = patches_pred.readlines()[1:]
+
+        for line in patches_pred_lines:
+            line_items = line[:-1].split(",")
+            line_class = line_items[2]
+            line_conf = float(line_items[3])
+            if (line_class in classes) and (line_conf >
+                                            conf_thresholds[line_class]):
```
Comment (jasonwei20) [inline code/utils_evaluation.py]: why is this on a separate line

File: code/utils_model.py [hunk 11]
```diff
@@ -1,479 +1,509 @@
...
-        return im.rotate(degree)
-
-
-# instantiate the model
-def create_model(num_layers, pretrain):
-    assert num_layers in [18, 24, 50, 101, 152]
-    architecture = "resnet{}".format(num_layers)
-    model_constructor = getattr(torchvision.models, architecture)
+
+def calculate_confusion_matrix(all_labels: np.ndarray,
```
Comment (jasonwei20) [inline code/utils_model.py]: have you tested it for a large number? the reason I had the `10000` stuff is because it took too long to update after having like 100k samples

File: code/utils_model.py [hunk 12]
```diff
@@ -1,479 +1,509 @@
...
 ###########################################
 #          MAIN TRAIN FUNCTION            #
 ###########################################
 
 
-# helper function for training resnet
-def train_helper(model, dataloaders, device, dataset_sizes, criterion,
-                 optimizer, scheduler, num_epochs, save_interval, writer):
-
+def train_helper(model: torchvision.models.resnet.ResNet,
```
Comment (jasonwei20) [inline code/utils_model.py]: i've never used type enforcement in python, but it this is what it is, then i love it

File: code/utils_model.py [hunk 13]
```diff
@@ -1,479 +1,509 @@
...
+    if config.args.resume_checkpoint:
+        ckpt = torch.load(f=config.resume_checkpoint_path)
+        model.load_state_dict(state_dict=ckpt["model_state_dict"])
+        optimizer.load_state_dict(state_dict=ckpt["optimizer_state_dict"])
+        scheduler.load_state_dict(state_dict=ckpt["scheduler_state_dict"])
+        epoch = ckpt["epoch"]
+        print(f"model loaded from {config.resume_checkpoint_path}")
     else:
-        model = create_model(num_layers, pretrain)
+        epoch = 0
```
Comment (jasonwei20) [inline code/utils_model.py]: what happened to created a model with pretraining?

File: code/utils_processing.py [hunk 14]
```diff
@@ -1,210 +1,160 @@
...
-                num_purple_squares += 1
-    if num_purple_squares > 100:
-        return True
-    return False
+    Returns:
+        A boolean representing whether the image is purple or not.
+    """
+    pooled = block_reduce(image=crop,
+                          block_size=(crop.shape[0] // 15, crop.shape[1] // 15,
+                                      1),
```
Comment (jasonwei20) [inline code/utils_processing.py]: why is this one a new line

File: code/utils_processing.py [hunk 15]
```diff
@@ -1,210 +1,160 @@
...
+                          func=np.average)
+
+    return pooled[(pooled[..., 0] > pooled[..., 1] - 10)
```
Comment (jasonwei20) [inline code/utils_processing.py]: is there a more clear way to write this?

File: code/utils_model.py (hunk 11 above)
Comment (JosephDiPalma) [reply code/utils_model.py]: @ntomita @jasonwei20 This commit allocates tensors at the beginning of the training/evaluation loop which store the output.  The confusion matrix is calculated once on the 1d predicted and actual class values at the end of training/evaluation.

File: code/utils_model.py (hunk 12 above)
Comment (JosephDiPalma) [reply code/utils_model.py]: In typical Python fashion, the types are never enforced, but an IDE will warn you if things are the wrong type.

File: code/utils_model.py (hunk 13 above)
Comment (JosephDiPalma) [reply code/utils_model.py]: @jasonwei20 The create_model() function now checks if the pretraining is needed instead of passing in the pretrain argument.

File: code/utils_model.py [hunk 16]
```diff
@@ -1,479 +1,509 @@
...
+                                     device=config.device)
+    val_all_labels = torch.empty(size=(dataset_sizes["val"], ),
+                                 dtype=torch.long,
+                                 device=config.device)
+    val_all_predicts = torch.empty(size=(dataset_sizes["val"], ),
+                                   dtype=torch.long,
+                                   device=config.device)
+
+    # Train for specified number of epochs.
+    for epoch in range(num_epochs, config.args.num_epochs):
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: @ntomita That's correct.  That's why I set epoch = 0 (utils_model line 370) if we don't want to resume from a checkpoint.

File: code/utils_model.py [hunk 17]
```diff
@@ -1,479 +1,509 @@
...
+                                   device=config.device)
+    train_all_predicts = torch.empty(size=(dataset_sizes["train"], ),
+                                     dtype=torch.long,
+                                     device=config.device)
+    val_all_labels = torch.empty(size=(dataset_sizes["val"], ),
+                                 dtype=torch.long,
+                                 device=config.device)
+    val_all_predicts = torch.empty(size=(dataset_sizes["val"], ),
+                                   dtype=torch.long,
+                                   device=config.device)
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: @ntomita I didn't think about using CPU memory.  I didn't want to do a transfer back to CPU memory from the GPU.  This probably should be done on the CPU to avoid memory usage on the GPU.

File: code/run_all.py (hunk 6 above)
Comment (JosephDiPalma) [reply code/run_all.py]: @jasonwei20 @ntomita This was left from my testing on Windows.  It should be removed for now.

File: code/config.py (hunk 8 above)
Comment (ntomita) [reply code/config.py]: To be honest I don't use pathlib by myself but after some research i found that it's recommended for Python 3.6+ programming as it's more object oriented approach. Well, technically no practical difference for now but it maybe good for us to use it since we have set the dependency of py3.6.

File: code/utils_processing.py [hunk 18]
```diff
@@ -214,132 +164,210 @@ def gen_val_patches(input_folder, output_folder, overlap_factor):
...
+    Duplicate the underrepresented classes to balance class distributions.
 
+    Args:
+        image_paths: Image paths to check for balance.
+        n: Desired number of images.
+    """
     num_dupls = n - len(image_paths)
 
-    print('balancing', basefolder(image_paths[0]), 'by duplicating', num_dupls)
+    print(f"balancing {image_paths[0].parent} " f"by duplicating {num_dupls}")
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: @ntomita I'll fix this.  This is a mistake from the PyCharm formatter.

File: code/config.py (hunk 8 above)
Comment (JosephDiPalma) [reply code/config.py]: @jasonwei20 This is the first time I used pathlib.  I used it since it is now the suggested way to deal with paths in Python.  It removes the need for the messy OS module.

File: code/config.py (hunk 9 above)
Comment (JosephDiPalma) [reply code/config.py]: @jasonwei20 @ntomita I will change this to a tuple since it should be immutable.

File: code/utils_model.py (hunk 11 above)
Comment (JosephDiPalma) [reply code/utils_model.py]: @jasonwei20 @ntomita We also cannot use the 10000 indexing version since that will return an incorrect confusion matrix when dealing with more than 10000 inputs.

File: code/utils_processing.py [hunk 19]
```diff
@@ -214,132 +164,210 @@ def gen_val_patches(input_folder, output_folder, overlap_factor):
...
+        # Put the (x, y) coordinates in the input queue.
+        for xy in itertools.product(range(0, x_steps * step_size, step_size),
+                                    range(0, y_steps * step_size, step_size)):
+            in_queue.put(obj=xy)
+
+        # Store num_workers None values so the processes exit when not enough jobs left.
+        for __ in range(config.args.num_workers):
+            in_queue.put(obj=None)
+
+        num_patches += sum(
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: @ntomita That's right.  I forgot to fix it earlier.

File: code/utils_model.py [hunk 20]
```diff
@@ -1,479 +1,509 @@
...
+
+    cm = pd.crosstab(index=pd.Series(pd.Categorical(
+        pd.Series(all_labels).replace(remap_classes),
+        categories=config.classes),
+                                     name="Actual"),
+                     columns=pd.Series(pd.Categorical(
+                         pd.Series(all_predicts).replace(remap_classes),
+                         categories=config.classes),
+                                       name="Predicted"),
+                     normalize="index")
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: @ntomita I'll fix the formatting.  The code formatting doesn't always work well with pandas.

File: code/compute_stats.py (hunk 3 above)
Comment (JosephDiPalma) [reply code/compute_stats.py]: @ntomita @jasonwei20 That's just unclearly explained.  I don't to run it at the patch level due to the overlap, but the idea is that the WSI are actually crops or tissues rather than the full WSI.

File: code/4_test.py (hunk 2 above)
Comment (JosephDiPalma) [reply code/4_test.py]: Added parameters back in newest commit.

File: code/config.py (hunk 4 above)
Comment (JosephDiPalma) [reply code/config.py]: Changed back in newest commit.

File: code/config.py (hunk 5 above)
Comment (JosephDiPalma) [reply code/config.py]: Changed back in newest commit.

File: code/3_train.py (hunk 1 above)
Comment (JosephDiPalma) [reply code/3_train.py]: Parameters added back in.  Now the `utils_*.py` files don't have access to the `config.py` file.

File: code/config.py (hunk 9 above)
Comment (JosephDiPalma) [reply code/config.py]: Changed to tuple in newest commit.

File: code/utils_processing.py (hunk 14 above)
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed to be more readable in newest commit.

File: code/utils_processing.py (hunk 15 above)
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed to be more readable in newest commit.

File: code/utils_processing.py [hunk 21]
```diff
@@ -1,210 +1,160 @@
...
+    image_paths = get_image_paths(folder=folder)
+
     file_size = 0
     for image_path in image_paths:
-        file_size += os.path.getsize(image_path)
-        # print(image_path, os.path.getsize(image_path))
-    file_size_mb = file_size / 1000.0 / 1000.0
-    return file_size_mb, len(image_paths)
+        file_size += image_path.stat().st_size
+    return (file_size / 1000.0 / 1000.0), len(image_paths)
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed in newest commit.

File: code/utils_processing.py [hunk 22]
```diff
@@ -1,210 +1,160 @@
...
-        overlap_factor = max(1, math.pow(overlap_factor,
-                                         1.5))  # empircal guessing?
+            folder=subfolder)
+
+        # Each image is 13KB = 0.013MB.
+        overlap_factor = max(
+            1.0,
+            math.pow(
+                math.sqrt(desired_crops_per_class / (subfolder_size / 0.013)),
+                1.5))
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: Added back in comment.

File: code/utils_processing.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed in newest commit.

File: code/utils_processing.py [hunk 23]
```diff
@@ -214,132 +164,210 @@ def gen_val_patches(input_folder, output_folder, overlap_factor):
...
-                    basename(image_path).split('_')[-2:])
-        os.system(" ".join(['cp', image_path, dupl_path]))
+        image_path = image_paths[i % len(image_paths)]
 
+        copyfile(src=image_path,
+                 dst=Path(
+                     image_path.parent,
+                     f"{'_'.join(image_path.name.split('_')[:-2])}dup"
+                     f"{(i // len(image_paths)) + 2}_"
+                     f"{'_'.join(image_path.name.split('_')[-2:])}"))
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed in newest commit.

File: code/utils_processing.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_processing.py]: Fixed in newest commit.

File: code/utils_model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: Fixed in newest commit.

File: code/utils_model.py [hunk 24]
```diff
@@ -1,479 +1,509 @@
...
-                train_preds == train_labels.data)
-            train_conf_matrix, train_all_labels, train_all_predicts = update_confusion_matrix(
-                train_all_labels, train_all_predicts, train_labels.data,
-                train_preds, config.num_classes)
+                train_preds == train_labels.data, dtype=torch.double)
+
+            train_all_labels[idx * config.args.batch_size:(idx + 1) *
+                             config.args.batch_size] = train_labels.data
+            train_all_predicts[idx * config.args.batch_size:(idx + 1) *
+                               config.args.batch_size] = train_preds
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: Fixed in newest commit.

File: code/utils_model.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: Tensors now stored in CPU memory.

File: code/utils_split.py [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+    Main function for splitting data. Note that we want the
+    validation and test sets to be balanced.
+    """
+    # Based on whether we want to move or keep the files.
+    head = shutil.copyfile if config.args.keep_orig_copy else shutil.move
+
+    # Create folders.
+    for f in [
+            config.args.wsi_train, config.args.wsi_val, config.args.wsi_test
+    ]:
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_split.py]: Converted to tuple.

File: code/compute_stats.py (hunk 3 above)
Comment (jasonwei2020) [reply code/compute_stats.py]: > The idea is that the WSI are actually crops or tissues rather than the full WSI.
How is the reader supposed to know this? I only ask because this is one of the cases where neural nets "fail silently", the performance will be not too great but it still "works" and so often the user doesn't know why the performance isn't better.

File: code/utils_model.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jasonwei2020) [reply code/utils_model.py]: Can you name it appropriately? Like `start_epoch` maybe.

File: README.md [hunk 26]
```diff
@@ -1,38 +1,81 @@
...
+- Replace [os](https://docs.python.org/3.6/library/os.html) with
+[pathlib](https://docs.python.org/3.6/library/pathlib.html) module.
+- Added [argparse](https://docs.python.org/3.6/library/argparse.html) support.
+- Rewrote patch processing code using vector operations.
+- Fully documented all functions and methods with docstrings and type hints.
+- Compute image means and standard deviations for normalization online.
+- Increased efficiency of code by rewriting inefficient code.
+
+## Known Issues and Limitations
+- Patch creation code doesn't find any patches in images that are slightly larger
```
Comment (jasonwei2020) [inline README.md]: Not clear to me what this means

File: README.md [hunk 27]
```diff
@@ -1,38 +1,81 @@
...
+than the specified patch size.
+- Only 1 GPU supported.
+- Should work, but not tested on Windows.
+- In cases where no crops are found for an image, empty directories are created.
+Current workaround uses `try` and `except` statements to catch errors.
+- Image reading code expects colors to be in the RGB space. Current workaround
+is to keep first 3 channels.
```
Comment (jasonwei2020) [inline README.md]: Why's this on a new line

File: README.md [hunk 28]
```diff
@@ -1,38 +1,81 @@
...
 # Usage
 
 Take a look at `code/config.py` before you begin to get a feel for what parameters can be changed.
 
 ## 1. Train-Val-Test Split:
 
-Splits the data into a validation and test set. Default validation whole-slide images (WSI) per class is 15 and test images per class is 30. You can change these numbers by changing `val_wsi_per_class` and `test_wsi_per_class` in `code/config.py`. You can skip this step if you did a custom split (for example, you need to split by patients).
+Splits the data into a validation and test set. 
+Default validation whole-slide images (WSI) per class is 20 and test images per class is 30. 
+You can change these numbers by changing the `--val-wsi-per-class` and \
```
Comment (jasonwei2020) [inline README.md]: Should these be in snake case? I.e., --val_wsi_per_class

File: README.md [hunk 29]
```diff
@@ -145,11 +229,10 @@ See `code/z_preprocessing` for some code to convert images from svs into jpg. Th
 # Still not working? Consider the following...
 
 - Ask a pathologist to look at your visualizations.
-- Make your own heuristic for aggregating patch predictions to get WSI classification. Often, a slide thats 20% abnormal and 80% normal should be classified as abnormal.
+- Make your own heuristic for aggregating patch predictions to find WSI classification. Often, a slide thats 20% abnormal and 80% normal should be classified as abnormal.
```
Comment (jasonwei2020) [inline README.md]: "to determine the WSI-level classification"

File: code/2_process_patches.py [hunk 30]
```diff
@@ -1,31 +1,37 @@
...
+gen_train_patches(config.args.wsi_train, config.train_patches,
+                  config.args.num_train_per_class)
+print("----- Finished generating training patches -----")
+print("----- Balancing the training patches -----")
+# Balance the training patches.
+balance_classes(config.train_patches)
+print("----- Finished balancing the training patches -----")
+print("----- Generating validation patches -----")
+# Generate validation patches.
+gen_val_patches(config.args.wsi_val, config.val_patches, overlap_factor=1.5)
```
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/2_process_patches.py]: Fixed in newest commit.

File: README.md (hunk 26 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 When running the processing code on images that were around 300x300 pixels, no patches were found even though we probably should find some.

File: README.md (hunk 28 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 The flags are currently expected to be in the `--val-wsi-per-class` form.

File: README.md (hunk 27 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 This is a formatting mistake.  I'll fix and commit.

File: code/compute_stats.py (hunk 3 above)
Comment (JosephDiPalma) [reply code/compute_stats.py]: @jasonwei20 I added in a note at the top of the README.  I'll commit the fix.

File: README.md (hunk 29 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 Fixed in newest commit.

File: code/utils_model.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (JosephDiPalma) [reply code/utils_model.py]: @jasonwei20 Fixed in newest commit.

File: code/run_all.py (hunk 6 above)
Comment (JosephDiPalma) [reply code/run_all.py]: @jasonwei20 @ntomita I ended up leaving the file in because the user cannot change the `argparse` flags if they use `run_all.sh`.

File: README.md [hunk 31]
```diff
@@ -1,38 +1,77 @@
...
 
 
 ![alt text](figures/figure-2-color.jpeg)
 
-## Dependencies
-- [Python 3.6](https://www.anaconda.com/download/#macos)
-- [PyTorch](https://pytorch.org/)
-- [PIL](https://pillow.readthedocs.io/en/5.3.x/)
+## Changes in this Release
+- Use Python [multiprocessing](https://docs.python.org/3.6/library/multiprocessing.html)
```
Comment (jasonwei20) [inline README.md]: This shouldn't be so far up in the readme. Can you also copy this into the PR?

File: README.md [hunk 32]
```diff
@@ -1,38 +1,77 @@
...
+- [SciPy](https://www.scipy.org/)
+- [NVIDIA GPU](https://www.nvidia.com/en-us/)
+- [Ubuntu](https://ubuntu.com/)
 
 # Usage
 
 Take a look at `code/config.py` before you begin to get a feel for what parameters can be changed.
 
+## Note on Expected WSI Input
+We expect the WSI input to be at the crop or tissue level, rather than the entire WSI. While the code will still work if entire WSIs are provided, the performance will likely be bad.
```
Comment (jasonwei20) [inline README.md]: I still don't get this. I used WSI for 3 of my projects, and they work. Why should the performance be bad on them?

File: README.md [hunk 33]
```diff
@@ -1,38 +1,77 @@
...
+You can change these numbers by changing the `--val-wsi-per-class` and \
+`--test-wsi-per-class` flags at runtime. 
+You can skip this step if you did a custom split (for example, you need to split by patients).
 
 ```
 python code/1_split.py
 ```
 
-If you do not want to duplicate the data, set `keep_orig_copy` in `code/config.py` to `False`.
+If you do not want to duplicate the data, append `--keep-orig-copy False` to the above command.
```
Comment (jasonwei20) [inline README.md]: Why aren't the argparse fields in snake case to begin with?

File: code/run_all.py (hunk 6 above)
Comment (jasonwei20) [reply code/run_all.py]: Cant they just edit the run_all file? This duplicated code is hard to maintain though, and in general, duplicated code is bad.

File: code/utils_split.py [hunk 34]
```diff
@@ -0,0 +1,139 @@
+"""
+DeepSlide
+Splits the data into training, validation, and testing sets.
+
+Authors: Jason Wei, Behnaz Abdollahi, Saeed Hassanpour
+
+Last Modified: November 30, 2019 (Joseph DiPalma)
```
Comment (jasonwei20) [inline code/utils_split.py]: This `Last Modified` is probably too hard to maintain and i think it should be removed from all files. 
This was before I knew `git blame` existed.

File: README.md (hunk 26 above)
Comment (jasonwei20) [reply README.md]: Can you show me an example? Many of the patches that I used were small and they all worked.

File: README.md (hunk 28 above)
Comment (jasonwei20) [reply README.md]: See my other comment

File: code/config.py [hunk 35]
```diff
@@ -1,120 +1,408 @@
...
+# Must contain subfolders of images labelled by class.
+# If your two classes are 'a' and 'n', you must have a/*.jpg with the images in class a and
+# n/*.jpg with the images in class n.
+parser.add_argument(
+    "--all-wsi",
+    type=Path,
+    default=Path("all_wsi"),
+    help="Location of the WSI organized in subfolders by class")
+# For splitting into validation set.
+parser.add_argument("--val-wsi-per-class",
```
Comment (jasonwei20) [inline code/config.py]: You can just changed this to `--val_wsi_per_class`.
It should be in snake case unless you have a clear reason why it shouldnt

**Comment by jasonwei20** [comment]
@JosephDiPalma 

From the first review:
-  What test cases have you run? Since many projects in our lab use this code, I would need to see an end-to-end test with a folder of images work (just show me the logs and the final accuracy is reasonable).
- At this point, do you want to be the maintainer of this repository?

File: README.md (hunk 32 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 The idea of this comment is that you need annotations instead of labels for the entire slide.  When running this network on WSIs that only have slide-level labels, the performance was terrible since the majority class was normal.

File: code/utils_split.py (hunk 34 above)
Comment (JosephDiPalma) [reply code/utils_split.py]: @jasonwei20 I didn't know about `git blame`.  It's a much better solution.

File: README.md (hunk 26 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 I realized my problem.  I used an overlap factor of 5 even though I should have used 3.

File: README.md (hunk 28 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 I'll switch the config file to snake case.

File: README.md (hunk 33 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 I'll switch the config file to snake case.

File: code/config.py (hunk 35 above)
Comment (JosephDiPalma) [reply code/config.py]: @jasonwei20 I'll switch to snake case.

File: README.md (hunk 31 above)
Comment (JosephDiPalma) [reply README.md]: @jasonwei20 Fixed in newest commit.

File: README.md (hunk 32 above)
Comment (ntomita) [reply README.md]: @JosephDiPalma I see your point but to be honest I don't like this sentence because 1. not sure if you can generalize that and 2. it's the first item users see in this readme, specifically "the performance will likely be bad" giving negative impression. Could you rephrase in non-negative way, and I think this should be better placed under Limitation section.

File: README.md [hunk 36]
```diff
@@ -1,38 +1,77 @@
...
+This repository is a sliding window framework for classification of high resolution whole-slide images, often called microscopy or histopathology images. This is also the code for the paper [Pathologist-level Classification of Histologic Patterns on Resected Lung Adenocarcinoma Slides with Deep Neural Networks](https://www.nature.com/articles/s41598-019-40041-7). For a practical guide and implementation tips, see the Medium post [Classification of Histopathology Images with Deep Learning: A Practical Guide](https://medium.com/health-data-science/classification-of-histopathology-images-with-deep-learning-a-practical-guide-2e3ffd6d59c5). 
 
 
 ![alt text](figures/figure-2-color.jpeg)
 
-## Dependencies
-- [Python 3.6](https://www.anaconda.com/download/#macos)
-- [PyTorch](https://pytorch.org/)
-- [PIL](https://pillow.readthedocs.io/en/5.3.x/)
+## Changes in this Release
```
Comment (ntomita) [inline README.md]: We should put a release tag for this commit and then this `Changes in this Release` should be placed there to describe the release.

File: README.md [hunk 37]
```diff
@@ -1,38 +1,77 @@
...
+to speed up patch creation.
+- Replace [os](https://docs.python.org/3.6/library/os.html) with
+[pathlib](https://docs.python.org/3.6/library/pathlib.html) module.
+- Added [argparse](https://docs.python.org/3.6/library/argparse.html) support.
+- Rewrote patch processing code using vector operations.
+- Fully documented all functions and methods with docstrings and type hints.
+- Compute image means and standard deviations for normalization online.
+- Increased efficiency of code by rewriting inefficient code.
+
+## Known Issues and Limitations
```
Comment (ntomita) [inline README.md]: Can we move this to before `Still not working?` section?

File: code/run_all.py (hunk 6 above)
Comment (ntomita) [reply code/run_all.py]: @JosephDiPalma Duplicated code is my main concern here. Can we actually call each numbered file with args here rather than reimplementing each process, maybe by using subprocess or os.system call?
